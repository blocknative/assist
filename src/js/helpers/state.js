export const initialState = {
  version: null,
  validApiKey: 'unknown',
  supportedNetwork: 'unknown',
  config: undefined,
  userAgent: null,
  mobileDevice: null,
  validBrowser: null,
  legacyWeb3: null,
  modernWeb3: null,
  web3Version: null,
  web3Instance: null,
  currentProvider: null,
  web3Wallet: null,
  legacyWallet: null,
  modernWallet: null,
  accessToAccounts: null,
  accountAddress: null,
  walletLoggedIn: null,
  walletEnabled: null,
  walletEnableCalled: null,
  walletEnableCanceled: null,
  accountBalance: null,
  minimumBalance: null,
  correctNetwork: null,
  userCurrentNetworkId: null,
  socket: null,
  pendingSocketConnection: null,
  socketConnection: null,
  transactionQueue: [],
  transactionAwaitingApproval: false,
  iframe: null,
  iframeDocument: null,
  iframeWindow: null,
  connectionId: null
}

export let state = Object.assign({}, initialState)

export function updateState(newState) {
  state = Object.assign({}, state, newState)
}

export function filteredState() {
  const {
    mobileDevice,
    validBrowser,
    currentProvider,
    web3Wallet,
    accessToAccounts,
    walletLoggedIn,
    walletEnabled,
    accountAddress,
    accountBalance,
    minimumBalance,
    userCurrentNetworkId,
    correctNetwork
  } = state

  return {
    mobileDevice,
    validBrowser,
    currentProvider,
    web3Wallet,
    accessToAccounts,
    walletLoggedIn,
    walletEnabled,
    accountAddress,
    accountBalance,
    minimumBalance,
    userCurrentNetworkId,
    correctNetwork
  }
}
