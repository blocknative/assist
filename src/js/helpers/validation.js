/* eslint-disable import/prefer-default-export */
import ow from 'ow'

const desktopPosition = ow.optional.string.is(
  s =>
    s === 'bottomLeft' ||
    s === 'bottomRight' ||
    s === 'topRight' ||
    s === 'topLeft'
)

const mobilePosition = ow.optional.string.is(s => s === 'bottom' || s === 'top')

export function validateConfig(config) {
  ow(
    config,
    'config',
    ow.object.exactShape({
      networkId: ow.number,
      dappId: ow.string,
      web3: ow.optional.object,
      mobileBlocked: ow.optional.boolean,
      minimumBalance: ow.optional.number,
      headlessMode: ow.optional.boolean,
      messages: ow.optional.object.exactShape({
        txRequest: ow.optional.function,
        txSent: ow.optional.function,
        txPending: ow.optional.function,
        txSendFail: ow.optional.function,
        txStall: ow.optional.function,
        txFailed: ow.optional.function,
        nsfFail: ow.optional.function,
        txRepeat: ow.optional.function,
        txAwaitingApproval: ow.optional.function,
        txConfirmReminder: ow.optional.function,
        txConfirmed: ow.optional.function,
        txSpeedUp: ow.optional.function
      }),
      images: ow.optional.object.exactShape({
        welcome: ow.optional.object.exactShape({
          src: ow.string,
          srcset: ow.string
        }),
        complete: ow.optional.object.exactShape({
          src: ow.string,
          srcset: ow.string
        })
      }),
      style: ow.optional.object.exactShape({
        darkMode: ow.optional.boolean,
        notificationsPosition: ow.any(
          desktopPosition,
          ow.optional.object.exactShape({
            mobile: mobilePosition,
            desktop: desktopPosition
          })
        ),
        css: ow.optional.string
      }),
      truffleContract: ow.optional.boolean
    })
  )
}
