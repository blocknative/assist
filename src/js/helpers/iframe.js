import { updateState, state } from './state'
import { positionElement } from '../views/dom'

import darkModeStyles from '../../css/dark-mode.css'

export function createIframe(browserDocument, assistStyles, style = {}) {
  const { darkMode, css } = style

  const initialIframeContent = `
    <html>
      <head>
        <style>
          ${assistStyles}
        </style>
        <style>
          ${darkMode ? darkModeStyles : ''}
        </style>
        <style>
          ${css || ''}
        </style>
      </head>
      <body></body>
    </html>
  `

  const iframe = positionElement(browserDocument.createElement('iframe'))
  browserDocument.body.appendChild(iframe)
  iframe.style.position = 'fixed'
  iframe.style.height = '0'
  iframe.style.width = '0'
  iframe.style.border = 'none'
  iframe.style.pointerEvents = 'none'
  iframe.style.overflow = 'hidden'
  const iWindow = iframe.contentWindow
  const iDocument = iWindow.document
  iDocument.open()
  iDocument.write(initialIframeContent)
  iDocument.close()

  updateState({ iframe, iframeDocument: iDocument, iframeWindow: iWindow })
}

export function showIframe() {
  state.iframe.style['z-index'] = 999
  state.iframe.style.pointerEvents = 'all'
}

export function hideIframe() {
  state.iframe.style['z-index'] = 'initial'
  state.iframe.style.pointerEvents = 'none'
}

export function resizeIframe({ height, width, transitionHeight }) {
  if (transitionHeight) {
    state.iframe.style.transition = 'height 200ms ease-in-out'
  } else {
    state.iframe.style.transition = 'initial'
  }

  state.iframe.style.height = `${height}px`
  state.iframe.style.width = `${width}px`
}
