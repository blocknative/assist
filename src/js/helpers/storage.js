import { state, updateState } from './state'

export function getItem(item) {
  let storageItem
  try {
    storageItem = window.localStorage && window.localStorage.getItem(item)
  } catch (errorObj) {
    return 'null'
  }
  return storageItem
}

export function storeItem(item, value) {
  try {
    window.localStorage && window.localStorage.setItem(item, value)
  } catch (errorObj) {
    return 'null'
  }
  return 'success'
}

export function removeItem(item) {
  try {
    window.localStorage && window.localStorage.removeItem(item)
  } catch (errorObj) {
    return 'null'
  }
  return 'success'
}

export function storeTransactionQueue() {
  const { transactionQueue } = state
  if (transactionQueue.length > 0) {
    const pendingTransactions = transactionQueue.filter(
      txObj =>
        txObj.transaction.status === 'approved' ||
        txObj.transaction.status === 'pending'
    )
    storeItem('transactionQueue', JSON.stringify(pendingTransactions))
  }
}

export function getTransactionQueueFromStorage() {
  const transactionQueue = getItem('transactionQueue')
  if (transactionQueue) {
    const parsedQueue = JSON.parse(transactionQueue)
    const filteredQueue = parsedQueue.filter(
      txObj => Date.now() - txObj.transaction.startTime < 150000
    )
    updateState({
      transactionQueue: [...filteredQueue, ...state.transactionQueue]
    })
    removeItem('transactionQueue')
  }
}
