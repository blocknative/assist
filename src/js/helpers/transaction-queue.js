import { state, updateState } from './state'

export function addTransactionToQueue(txObject) {
  const { transactionQueue } = state
  return [...transactionQueue, txObject]
}

export function removeTransactionFromQueue(id) {
  const { transactionQueue } = state
  const newQueueState = transactionQueue.filter(
    txObj => txObj.transaction.id !== id
  )

  updateState({ transactionQueue: newQueueState })
}

export function updateTransactionInQueue(id, update) {
  const txObject = getTxObjFromQueue(id)
  txObject.transaction = Object.assign(txObject.transaction, update)

  return txObject
}

export function getTxObjFromQueue(id) {
  const { transactionQueue } = state
  return transactionQueue.find(txObj => txObj.transaction.id === id)
}

export function isDuplicateTransaction({ value, to }) {
  const { transactionQueue } = state

  return Boolean(
    transactionQueue.find(
      txObj => txObj.transaction.value === value && txObj.transaction.to === to
    )
  )
}

export function getTransactionsAwaitingApproval() {
  const { transactionQueue } = state
  return transactionQueue.filter(
    txObj => txObj.transaction.status === 'awaitingApproval'
  )
}

export function isTransactionAwaitingApproval(id) {
  const txObj = getTxObjFromQueue(id)
  return txObj && txObj.transaction.status === 'awaitingApproval'
}
