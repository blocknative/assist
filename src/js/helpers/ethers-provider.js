import { state } from './state'
import { networkName } from './utilities'

let provider

export const getEthersProvider = () => {
  if (!provider) {
    const { ethers } = state.config
    if (typeof window.ethereum !== 'undefined') {
      provider = new ethers.providers.Web3Provider(window.ethereum)
    } else if (typeof window.web3 !== 'undefined') {
      provider = new ethers.providers.Web3Provider(window.web3.currentProvider)
    } else {
      provider = ethers.getDefaultProvider(networkName(state.config.networkId))
    }
  }

  return provider
}

let UncheckedSigner

export const getUncheckedSigner = () => {
  if (!UncheckedSigner) {
    const { ethers } = state.config
    const provider = getEthersProvider()
    class UncheckedJsonRpcSigner extends ethers.Signer {
      constructor(signer) {
        super()
        ethers.utils.defineReadOnly(this, 'signer', signer)
        ethers.utils.defineReadOnly(this, 'provider', signer.provider)
      }

      getAddress() {
        return this.signer.getAddress()
      }

      sendTransaction(transaction) {
        return this.signer.sendUncheckedTransaction(transaction).then(hash => ({
          hash,
          nonce: null,
          gasLimit: null,
          gasPrice: null,
          data: null,
          value: null,
          chainId: null,
          confirmations: 0,
          from: null,
          wait: confirmations =>
            this.provider.waitForTransaction(hash, confirmations)
        }))
      }
    }

    UncheckedSigner = new UncheckedJsonRpcSigner(provider.getSigner())
  }

  return UncheckedSigner
}
