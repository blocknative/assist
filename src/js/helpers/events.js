import { state } from './state'
import eventToUI from '../views/event-to-ui'
import { networkName, timeouts } from './utilities'
import { getTxObjFromQueue } from './transaction-queue'
import { openWebsocketConnection } from './websockets'
import { getItem } from './storage'

export function handleEvent(eventObj, clickHandlers) {
  const { eventCode, categoryCode } = eventObj
  const serverEvent =
    eventCode === 'txPending' ||
    eventCode === 'txConfirmed' ||
    eventCode === 'txFailed'

  // If not a server event then log it
  !serverEvent && lib.logEvent(eventObj)

  // If tx status is 'completed', UI has been already handled
  if (eventCode === 'txConfirmed' || eventCode === 'txConfirmedClient') {
    const txObj = getTxObjFromQueue(eventObj.transaction.id)

    if (txObj.transaction.status === 'completed') {
      return
    }
  }

  // Show UI
  if (state.config && !state.config.headlessMode) {
    eventToUI[categoryCode] &&
      eventToUI[categoryCode][eventCode] &&
      eventToUI[categoryCode][eventCode](eventObj, clickHandlers)
  }
}

// Create event log to be sent to server
export function createEventLog(eventObj) {
  const { dappId, networkId, headlessMode } = state.config
  const { userAgent, version } = state
  const newUser = getItem('_assist_newUser') === 'true'
  return JSON.stringify(
    Object.assign(
      {},
      {
        timeStamp: new Date(),
        dappId,
        version,
        userAgent,
        newUser,
        headlessMode,
        blockchain: {
          system: 'ethereum',
          network: networkName(networkId)
        }
      },
      eventObj
    )
  )
}

// Log events to server
export function logEvent(eventObj) {
  const eventLog = createEventLog(eventObj)
  const { socket, socketConnection } = state

  socketConnection && socket.send(eventLog)

  // Need to check if connection dropped
  // as we don't know until after we try to send a message
  setTimeout(() => {
    if (!state.socketConnection) {
      if (!state.pendingSocketConnection) {
        openWebsocketConnection()
      }

      setTimeout(() => {
        logEvent(eventObj)
      }, timeouts.checkSocketConnection)
    }
  }, timeouts.checkSocketConnection)
}

export const lib = {
  handleEvent,
  createEventLog,
  logEvent
}
