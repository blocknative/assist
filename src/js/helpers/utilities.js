import uuid from 'uuid/v4'
import { handleEvent } from './events'

// Nice time format
export function formatTime(number) {
  const time = new Date(number)
  return time.toLocaleString('en-US', {
    hour: 'numeric',
    minute: 'numeric',
    hour12: true
  })
}

export function timeString(time) {
  const seconds = Math.floor(time / 1000)
  return seconds >= 60 ? `${Math.floor(seconds / 60)} min` : `${seconds} sec`
}

export function capitalize(str) {
  const first = str.slice(0, 1)
  const rest = str.slice(1)
  return `${first.toUpperCase()}${rest}`
}

export function formatNumber(num) {
  const numString = String(num)
  if (numString.includes('+')) {
    const exponent = numString.split('+')[1]
    const precision = Number(exponent) + 1
    return num.toPrecision(precision)
  }
  return num
}

function last(arr) {
  return [...arr].reverse()[0]
}

function takeLast(arr) {
  // mutates original array
  return arr.splice(arr.length - 1, 1)[0]
}

export function first(arr) {
  return arr[0]
}

function takeFirst(arr) {
  // mutates original arr
  return arr.splice(0, 1)[0]
}

export function createTransactionId() {
  return uuid()
}

export function separateArgs(allArgs, argsLength) {
  const allArgsCopy = [...allArgs]
  const args = argsLength ? allArgsCopy.splice(0, argsLength) : []

  const inlineCustomMsgs =
    typeof last(allArgsCopy) === 'object' &&
    last(allArgsCopy).messages &&
    takeLast(allArgsCopy).messages

  const callback =
    typeof last(allArgsCopy) === 'function' && takeLast(allArgsCopy)

  const txObject =
    typeof first(allArgsCopy) === 'object' && first(allArgsCopy) !== null
      ? takeFirst(allArgsCopy)
      : {}

  const defaultBlock = first(allArgsCopy)

  return {
    callback,
    args,
    txObject,
    defaultBlock,
    inlineCustomMsgs
  }
}

export function getOverloadedMethodKeys(inputs) {
  return inputs.map(input => input.type).join(',')
}

export function assistLog(log) {
  console.log('Assist:', log) // eslint-disable-line no-console
}

export function extractMessageFromError(message) {
  const str = message.split('"message":')[1]
  return str.split('"')[1]
}

export function eventCodeToType(eventCode) {
  switch (eventCode) {
    case 'txRequest':
    case 'txPending':
    case 'txSent':
      return 'progress'
    case 'txSendFail':
    case 'txStall':
    case 'txFailed':
    case 'nsfFail':
    case 'txRepeat':
    case 'txAwaitingApproval':
    case 'txConfirmReminder':
      return 'failed'
    case 'txConfirmed':
    case 'txConfirmedClient':
      return 'complete'
    default:
      return undefined
  }
}

export function eventCodeToStep(eventCode) {
  switch (eventCode) {
    case 'mobileBlocked':
      return 'mobile'
    case 'browserFail':
      return 'browser'
    case 'welcomeUser':
      return 0
    case 'walletFail':
      return 1
    case 'walletLogin':
    case 'walletLoginEnable':
      return 2
    case 'walletEnable':
      return 3
    case 'networkFail':
    case 'contractQueryFail':
      return 4
    case 'nsfFail':
      return 5
    case 'newOnboardComplete':
      return 6
    default:
      return undefined
  }
}

export function networkName(id) {
  switch (Number(id)) {
    case 1:
      return 'main'
    case 3:
      return 'ropsten'
    case 4:
      return 'rinkeby'
    case 42:
      return 'kovan'
    case 'localhost':
      return 'localhost'
    default:
      return 'local'
  }
}

export const timeouts = {
  checkSocketConnection: 250,
  waitForResponse: 100,
  txConfirmReminder: 20000,
  txStall: 30000,
  changeUI: 305,
  localhostNetworkCheck: 300,
  removeElement: 300,
  endOfEventQueue: 0,
  hideElement: 200,
  showElement: 120,
  autoRemoveNotification: 4000,
  pollForReceipt: 1000
}

export function stepToImageKey(step) {
  switch (step) {
    case 0:
      return 'welcome'
    case 6:
      return 'complete'
    default:
      return null
  }
}

export function handleError(handlers = {}) {
  return errorObj => {
    const { callback, reject, resolve } = handlers

    if (callback) {
      callback(errorObj)
      resolve()

      return
    }

    reject(errorObj)
  }
}

export function handleWeb3Error(errorObj) {
  const { message } = errorObj
  handleEvent({
    eventCode: 'errorLog',
    categoryCode: 'web3',
    reason: message || errorObj
  })
}
