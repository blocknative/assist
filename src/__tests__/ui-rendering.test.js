/*
 * Check that UI is rendered correctly in response to events
 */

import { handleEvent } from '../js/helpers/events'
import { createIframe } from '../js/helpers/iframe'
import { state, updateState } from '../js/helpers/state'
import assistStyles from '../css/styles.css'

// Make a copy of the initial state
const initialState = Object.assign({}, state)

const mockBareTransaction = {
  id: 'cf7fc5a7-1498-419e-8bf9-654d06af5534',
  gas: '54707',
  gasPrice: '1000000000',
  value: '1848587500000000',
  to: '0x9b913956036a3462330b0642b20d3879ce68b450',
  from: '0x90f8bf6a479f320ead074411a4b0e7944ea8c9c1'
}

const mockBroadcastTransaction = Object.assign({}, mockBareTransaction, {
  nonce: 1235,
  startTime: 1262264000000,
  txSent: true,
  inTxPool: true
})

// Specify events to test here ([categoryCode, eventCode, txShouldBeBroadcasted])
// Setting txShouldBeBroadcasted to true asserts that an error should be thrown if
// 'nonce', 'startTime', 'txSent' or 'inTxPool' is missing from the passed transaction object
const eventsToTest = [
  ['activeTransaction', 'txRequest', false],
  ['activeTransaction', 'txPending', true],
  ['activeTransaction', 'txSent', true]
]

describe('ui-rendering', () => {
  describe('notificationsUI', () => {
    eventsToTest.forEach(eventInfo => {
      const [categoryCode, eventCode, txShouldBeBroadcasted] = eventInfo
      // Decide which transaction type to pass with the event
      const transaction = txShouldBeBroadcasted
        ? mockBroadcastTransaction
        : mockBareTransaction
      describe(`event ${categoryCode}-${eventCode}`, () => {
        test('should trigger correct DOM render', () => {
          handleEvent({
            categoryCode,
            eventCode,
            transaction
          })
          expect(state.iframeDocument.body.innerHTML).toMatchSnapshot()
        })
        if (txShouldBeBroadcasted) {
          test('should fail if transaction is missing attributes', () => {
            expect(() => {
              handleEvent({
                categoryCode,
                eventCode,
                transaction: mockBareTransaction
              })
            }).toThrow()
          })
        }
      })
    })
  })
})

// Create an iframe and reset state between each test
beforeEach(() => {
  updateState({ config: {} })
  createIframe(document, assistStyles)
})

afterEach(() => {
  updateState(Object.assign({}, initialState))
})
