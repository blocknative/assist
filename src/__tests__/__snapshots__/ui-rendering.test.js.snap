// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui-rendering notificationsUI event activeTransaction-txPending should trigger correct DOM render 1`] = `
"
    
  <div class=\\"\\" id=\\"blocknative-notifications\\" style=\\"transform: translate(600px); right: 0px; bottom: 0px;\\"><div class=\\"bn-notifications-scroll\\" style=\\"overflow-y: initial; overflow-x: initial; height: auto;\\"><ul class=\\"bn-notifications\\"><li class=\\"bn-notification bn-progress bn-txPending bn-cf7fc5a7-1498-419e-8bf9-654d06af5534 \\" style=\\"transform: translate(600px);\\">
		<span class=\\"bn-status-icon \\">
      <div class=\\"progress-tooltip \\">
				<div class=\\"progress-tooltip-inner\\">
					You will be notified when this transaction is completed.
				</div>
			</div>
		</span>
		<div class=\\"bn-notification-info\\">
			<p>Your transaction ID: 1235 has started</p>
			<p class=\\"bn-notification-meta\\">
				<a href=\\"#\\" class=\\"bn-timestamp\\">12:00 AM</a>
				<span class=\\"bn-duration\\"> - 
					<i class=\\"bn-clock\\"></i>
					<span class=\\"bn-duration-time\\">6 min</span>
				</span>
			</p>
		</div>
	</li></ul></div><a class=\\"\\" id=\\"bn-transaction-branding\\" href=\\"https://www.blocknative.com/\\" target=\\"_blank\\" style=\\"float: right;\\"></a></div>"
`;

exports[`ui-rendering notificationsUI event activeTransaction-txRequest should trigger correct DOM render 1`] = `
"
    
  <div class=\\"\\" id=\\"blocknative-notifications\\" style=\\"transform: translate(600px); right: 0px; bottom: 0px;\\"><div class=\\"bn-notifications-scroll\\" style=\\"overflow-y: initial; overflow-x: initial; height: auto;\\"><ul class=\\"bn-notifications\\"><li class=\\"bn-notification bn-progress bn-txRequest bn-cf7fc5a7-1498-419e-8bf9-654d06af5534 \\" style=\\"transform: translate(600px);\\">
		<span class=\\"bn-status-icon \\">
      <div class=\\"progress-tooltip \\">
				<div class=\\"progress-tooltip-inner\\">
					You will be notified when this transaction is completed.
				</div>
			</div>
		</span>
		<div class=\\"bn-notification-info\\">
			<p>Your transaction is waiting for you to confirm</p>
			<p class=\\"bn-notification-meta\\">
				<a href=\\"#\\" class=\\"bn-timestamp\\">12:00 AM</a>
				<span class=\\"bn-duration bn-duration-hidden\\"> - 
					<i class=\\"bn-clock\\"></i>
					<span class=\\"bn-duration-time\\">NaN sec</span>
				</span>
			</p>
		</div>
	</li></ul></div><a class=\\"\\" id=\\"bn-transaction-branding\\" href=\\"https://www.blocknative.com/\\" target=\\"_blank\\" style=\\"float: right;\\"></a></div>"
`;

exports[`ui-rendering notificationsUI event activeTransaction-txSent should trigger correct DOM render 1`] = `
"
    
  <div class=\\"\\" id=\\"blocknative-notifications\\" style=\\"transform: translate(600px); right: 0px; bottom: 0px;\\"><div class=\\"bn-notifications-scroll\\" style=\\"overflow-y: initial; overflow-x: initial; height: auto;\\"><ul class=\\"bn-notifications\\"><li class=\\"bn-notification bn-progress bn-txSent bn-cf7fc5a7-1498-419e-8bf9-654d06af5534 \\" style=\\"transform: translate(600px);\\">
		<span class=\\"bn-status-icon \\">
      <div class=\\"progress-tooltip \\">
				<div class=\\"progress-tooltip-inner\\">
					You will be notified when this transaction is completed.
				</div>
			</div>
		</span>
		<div class=\\"bn-notification-info\\">
			<p>Your transaction ID: 1235 has been sent to the network</p>
			<p class=\\"bn-notification-meta\\">
				<a href=\\"#\\" class=\\"bn-timestamp\\">12:00 AM</a>
				<span class=\\"bn-duration\\"> - 
					<i class=\\"bn-clock\\"></i>
					<span class=\\"bn-duration-time\\">6 min</span>
				</span>
			</p>
		</div>
	</li></ul></div><a class=\\"\\" id=\\"bn-transaction-branding\\" href=\\"https://www.blocknative.com/\\" target=\\"_blank\\" style=\\"float: right;\\"></a></div>"
`;
